pkg/modulewriter/tfwriter.go

```diff
@@ -459,40 +458,73 @@ func orderKeys[T any](settings map[string]T) []string {
 	return keys
 }
 
-func filterVarsByGraph(vars map[string]cty.Value, group config.DeploymentGroup, graph map[string][]config.ModConnection) map[string]cty.Value {
+func getUsedDeploymentVars(group config.DeploymentGroup, bp config.Blueprint) map[string]cty.Value {
 	// labels must always be written as a variable as it is implicitly added
 	groupInputs := map[string]bool{
 		"labels": true,
 	}
+
 	for _, mod := range group.Modules {
-		if connections, ok := graph[mod.ID]; ok {
-			for _, conn := range connections {
-				if conn.IsDeploymentKind() {
-					for _, v := range conn.GetSharedVariables() {
-						groupInputs[v] = true
-					}
-				}
-			}
+		for _, v := range config.GetUsedDeploymentVars(mod.Settings.AsObject()) {
+			groupInputs[v] = true
 		}
 	}
 
 	filteredVars := make(map[string]cty.Value)
-	for key, val := range vars {
+	for key, val := range bp.Vars.Items() {
 		if groupInputs[key] {
 			filteredVars[key] = val
 		}
 	}
 	return filteredVars
 }
 
-func getAllOutputs(group config.DeploymentGroup) []string {
-	outputs := make(map[string]bool)
+func substituteIgcReferences(mods []config.Module, igcRefs map[config.Reference]modulereader.VarInfo) []config.Module {
+	doctoredMods := make([]config.Module, len(mods))
+	for i, mod := range mods {
+		doctoredMods[i] = substituteIgcReferencesInModule(mod, igcRefs)
+	}
+	return doctoredMods
+}
+
+// Updates expressions in Module settings to use special IGC var name instead of the module reference
+func substituteIgcReferencesInModule(mod config.Module, igcRefs map[config.Reference]modulereader.VarInfo) config.Module {
+	v, _ := cty.Transform(mod.Settings.AsObject(), func(p cty.Path, v cty.Value) (cty.Value, error) {
+		e, is := config.IsExpressionValue(v)
+		if !is {
+			return v, nil
+		}
+		ue := e.RenderHclAsString()
+		for _, r := range e.References() {
+			oi, exists := igcRefs[r]
+			if !exists {
+				continue
+			}
+			s := fmt.Sprintf("module.%s.%s", r.Module, r.Name)
```

Comment:
This "last minute" change from a HCL module reference to a variable reference seems non-ideal. The issues:

1. There are points in the code where the reference appears to be an outwardly valid "intragroup" `module.x.y` reference. This can be a source of confusion for future development.
2. We're treating intergroup references as the ones to treat special rather than converting all Expression to correct HCL representations simultaneously.

That said, I see a comment at `RenderHclAsString` which suggests you see issues here, and different (worse) versions of this problem already exist. I'd like to merge this but discuss offline how you plan to address RenderHclAsString.